services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -201,44 +214,70 @@ private void initFileDescriptorMaps() {
     }
 
     @Override
+    public void run() {
+      if (request != null) {
+        handleReflectionRequest();
+      }
+    }
+
+    @Override
     public void onNext(ServerReflectionRequest request) {
-      initFileDescriptorMaps();
-      switch (request.getMessageRequestCase()) {
-        case FILE_BY_FILENAME:
-          getFileByName(request);
-          return;
-        case FILE_CONTAINING_SYMBOL:
-          getFileContainingSymbol(request);
-          return;
-        case FILE_CONTAINING_EXTENSION:
-          getFileByExtension(request);
-          return;
-        case ALL_EXTENSION_NUMBERS_OF_TYPE:
-          getAllExtensions(request);
-          return;
-        case LIST_SERVICES:
-          listServices(request);
-          return;
-        default:
-          sendErrorResponse(request, Status.UNIMPLEMENTED, "");
+      Preconditions.checkState(this.request == null);
+      this.request = request;
+      handleReflectionRequest();
+    }
+
+    private void handleReflectionRequest() {
+      if (serverCallStreamObserver.isReady()) {
+        initFileDescriptorMaps();
+        switch (request.getMessageRequestCase()) {
+          case FILE_BY_FILENAME:
+            getFileByName(request);
+            break;
+          case FILE_CONTAINING_SYMBOL:
+            getFileContainingSymbol(request);
+            break;
+          case FILE_CONTAINING_EXTENSION:
+            getFileByExtension(request);
+            break;
+          case ALL_EXTENSION_NUMBERS_OF_TYPE:
+            getAllExtensions(request);
+            break;
+          case LIST_SERVICES:
+            listServices(request);
+            break;
+          default:
+            sendErrorResponse(request, Status.UNIMPLEMENTED, "");
+        }
+        request = null;
+        if (closeAfterSend) {
+          serverCallStreamObserver.onCompleted();
+        } else {
+          serverCallStreamObserver.request(1);
+        }
       }
     }
 
     @Override
     public void onCompleted() {
-      responseObserver.onCompleted();
+      if (request != null) {
+        closeAfterSend = true;
+        handleReflectionRequest();
```

Comment:
No need to call this method and instead just wait for the `onReadyHandler` to be called. I guess it doesn't hurt though.